code/workspaces/infrastructure-api/src/config/routes.js

```diff
@@ -32,13 +32,14 @@ function configureRoutes(app) {
   app.put('/projects/:projectKey', permissionWrapper(PROJECTS_EDIT), projects.projectDocumentValidator(), projects.urlAndBodyProjectKeyMatchValidator(), ew(projects.createOrUpdateProject));
   app.delete('/projects/:projectKey', permissionWrapper(PROJECTS_DELETE), projects.actionWithKeyValidator(), ew(projects.deleteProjectByKey));
   app.get('/stacks', permissionWrapper(STACKS_LIST), stacks.listStacks);
-  app.get('/stacks/category/:category', permissionWrapper(STACKS_LIST), stacks.withCategoryValidator, stacks.listByCategory);
-  app.get('/stacks/mount/:mount', permissionWrapper(STORAGE_LIST), stacks.withMountValidator, stacks.listByMount);
-  app.get('/stacks/:name/isUnique', permissionWrapper(STACKS_LIST), stack.withNameValidator, names.isUnique);
+  app.get('/stacks/:projectKey', projectPermissionWrapper(STACKS_LIST), stacks.withProjectValidator, stacks.listByProject);
+  app.get('/stacks/:projectKey/category/:category', projectPermissionWrapper(STACKS_LIST), stacks.withCategoryValidator, stacks.listByCategory);
+  app.get('/stacks/:projectKey/mount/:mount', projectPermissionWrapper(STORAGE_LIST), stacks.withMountValidator, stacks.listByMount);
   app.get('/stack/id/:id', permissionWrapper(STACKS_OPEN), stack.withIdValidator, stack.getOneById);
-  app.get('/stack/name/:name', permissionWrapper(STACKS_CREATE), stack.withNameValidator, stack.getOneByName);
-  app.delete('/stack', permissionWrapper(STACKS_DELETE), stack.deleteStackValidator, stack.deleteStack);
-  app.post('/stack', permissionWrapper(STACKS_CREATE), stack.createStackValidator, stack.createStack);
+  app.get('/stack/:projectKey/name/:name', projectPermissionWrapper(STACKS_CREATE), stack.withNameValidator, stack.getOneByName);
+  app.get('/stack/:projectKey/:name/isUnique', projectPermissionWrapper(STACKS_LIST), stack.withNameValidator, names.isUnique);
+  app.post('/stack/:projectKey', projectPermissionWrapper(STACKS_CREATE), stack.createStackValidator, stack.createStack);
+  app.delete('/stack/:projectKey', projectPermissionWrapper(STACKS_DELETE), stack.deleteStackValidator, stack.deleteStack);
   app.post('/volume/query', permissionWrapper(STORAGE_LIST), volume.coreVolumeValidator, volume.queryVolume);
```

Comment:
Can you also wrap the volume routes with projectPermissionWrapper?

Reply:
I certainly can. Wasn't sure if you had made any changes to these lines yourself. Will include them in the volume services too.